File: python/src/herminas_kernel/permissions.py
from __future__ import annotations

from enum import Enum


class Role(str, Enum):
    """HermiNas' four RBAC roles (cahier des charges §7.1). Mirrors
    kernel/permissions/permissions.go and Rust's
    herminas_kernel::permissions::Role. Full RBAC middleware lands in
    M1.5/M7.1."""

    VIEWER = "viewer"
    ANALYST = "analyst"
    ENGINEER = "engineer"
    ADMIN = "admin"


_RANK = {
    Role.VIEWER: 0,
    Role.ANALYST: 1,
    Role.ENGINEER: 2,
    Role.ADMIN: 3,
}


def rank(role: Role) -> int:
    return _RANK[role]


def at_least(role: Role, minimum: Role) -> bool:
    return rank(role) >= rank(minimum)

File: rust/kernel/src/permissions.rs
/// HermiNas' four RBAC roles (cahier des charges §7.1). Mirrors
/// kernel/permissions/permissions.go and herminas_kernel.permissions
/// (Python). Full RBAC middleware lands in M1.5/M7.1.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Role {
    Viewer,
    Analyst,
    Engineer,
    Admin,
}

impl Role {
    pub fn rank(self) -> u8 {
        match self {
            Role::Viewer => 0,
            Role::Analyst => 1,
            Role::Engineer => 2,
            Role::Admin => 3,
        }
    }

    pub fn at_least(self, min: Role) -> bool {
        self.rank() >= min.rank()
    }

    pub fn all() -> [Role; 4] {
        [Role::Viewer, Role::Analyst, Role::Engineer, Role::Admin]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn at_least_respects_hierarchy() {
        assert!(Role::Admin.at_least(Role::Engineer));
        assert!(!Role::Viewer.at_least(Role::Analyst));
    }
}
